src/h3/mod.rs

```diff
@@ -1080,22 +1108,45 @@ impl Connection {
             return Ok((finished, Event::Finished));
         }
 
-        // Process DATAGRAMs
-        let mut d = [0; 8];
+        // Process DATAGRAMs.
+        let process_dgram = match self.dgram_poll_threshold {
+            None => true,
 
-        match conn.dgram_recv_peek(&mut d, 8) {
-            Ok(_) => {
-                let mut b = octets::Octets::with_slice(&d);
-                let flow_id = b.get_varint()?;
-                return Ok((flow_id, Event::Datagram));
-            },
+            Some(thresh) => self.dgram_poll_count < thresh,
+        };
 
-            Err(crate::Error::Done) => (),
+        if process_dgram {
+            let mut d = [0; 8];
 
-            Err(e) => return Err(Error::TransportError(e)),
-        };
+            match conn.dgram_recv_peek(&mut d, 8) {
+                Ok(_) => {
+                    if self.dgram_poll_threshold.is_some() {
+                        self.dgram_poll_count += 1;
+                    }
+
+                    let mut b = octets::Octets::with_slice(&d);
+                    let flow_id = b.get_varint()?;
+                    return Ok((flow_id, Event::Datagram));
+                },
+
+                Err(crate::Error::Done) => (),
+
+                Err(e) => return Err(Error::TransportError(e)),
+            };
+        }
 
         // Process HTTP/3 data from readable streams.
+        if let Some(thresh) = self.stream_poll_threshold {
+            if self.stream_poll_count > thresh {
```

Comment:
Should probably also move this inside the loop, to after `stream_poll_count` is incremented (see other comment), and change the condition to `>=`. Then we can also avoid returning `Done` here.